File: product/src/main/java/$group__/$modId__/client/gui/themes/IThemed.java
package $group__.$modId__.client.gui.themes;

import $group__.$modId__.utilities.concurrent.IMutatorImmutablizable;
import $group__.$modId__.logging.ILogging;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import org.apache.logging.log4j.Logger;

import javax.annotation.Nullable;
import java.util.Optional;

import static $group__.$modId__.utilities.helpers.specific.Throwables.rejectUnsupportedOperationIf;

@SideOnly(Side.CLIENT)
public interface IThemed<T extends ITheme<T>> {
	/* SECTION static methods */

	static <T extends ITheme<T>, L extends Logger> IThemed<T> of(@Nullable T theme, IMutatorImmutablizable<?, ?> mutator, ILogging<Logger> logging) { return theme == null ? GuiThemedNull.getInstance() : new GuiThemed<>(theme, mutator, logging); }


	/* SECTION getters & setters */

	@Nullable
	T getTheme();

	boolean trySetTheme(@Nullable T theme);


	default Optional<? extends T> tryGetTheme() { return Optional.ofNullable(getTheme()); }

	default void setTheme(@Nullable T theme) throws UnsupportedOperationException { rejectUnsupportedOperationIf(!trySetTheme(theme)); }
}

File: product/src/main/scala/$group__/$modId__/client/gui/themes/IThemed.scala
package $group__.$modId__.client.gui.themes

import java.util.Optional

import javax.annotation.Nullable
import org.apache.logging.log4j.Logger

@SideOnly(Side.CLIENT)
object IThemed {
	def of[T <: ITheme[T], L <: Logger](@Nullable theme: T, mutator: IMutatorImmutablizable[_, _], logging: ILogging[Logger]): IThemed[T] = {
		if (theme == null) GuiThemedNull.getInstance[T]
		else new GuiThemed(theme, mutator, logging)
	}
}

@SideOnly(Side.CLIENT)
trait IThemed[T <: ITheme[T]] {
	@Nullable
	def getTheme: T

	@throws[UnsupportedOperationException]
	def setTheme(@Nullable theme: T): Unit = rejectUnsupportedOperationIf(!trySetTheme(theme))

	def trySetTheme(@Nullable theme: T): Boolean

	def tryGetTheme: Optional[T] = Optional.ofNullable(getTheme)
}
